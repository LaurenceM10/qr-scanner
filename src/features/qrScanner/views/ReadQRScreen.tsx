import React, { useState } from 'react';
import {
  Text,
  View,
  Alert,
  StyleSheet,
  SafeAreaView,
  TouchableHighlight,
} from 'react-native';
import { BarCodeScanner, BarCodeScannerResult } from 'expo-barcode-scanner';

// Styles
import { theme } from 'app/theme';

// Utils/Hooks
import useScannerPermission from 'app/hooks/useScannerPermission';

// Redux
import { useAppDispatch } from 'app/store';
import { addQR } from 'features/qrScanner/slice';

function ReadQRScreen() {
  const dispatch = useAppDispatch();
  const hasPermission = useScannerPermission();
  const [scanned, setScanned] = useState(false);

  const handleBarCodeScanned = ({ data }: BarCodeScannerResult) => {
    setScanned(true);

    showConfirmAlert(`Do you want to save ${data}?`, () => {
      dispatch(addQR({ id: Date.now(), content: data }));
    });
  };

  const showConfirmAlert = (message: string, handleSave: () => void) => {
    Alert.alert('Scanned Successfully', message, [
      {
        text: 'Cancel',
        style: 'cancel',
      },
      { text: 'Save', onPress: handleSave },
    ]);
  };

  if (hasPermission === null) {
    return (
      <SafeAreaView style={styles.container}>
        <Text style={styles.permissionError}>
          Requesting for camera permission
        </Text>
      </SafeAreaView>
    );
  }

  if (!hasPermission) {
    return (
      <SafeAreaView style={styles.container}>
        <Text style={styles.permissionError}>No access to camera</Text>
      </SafeAreaView>
    );
  }

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.scannerContainer}>
        <BarCodeScanner
          onBarCodeScanned={scanned ? undefined : handleBarCodeScanned}
          style={StyleSheet.absoluteFillObject}
        />
        {scanned && (
          <View style={styles.scanButtonContainer}>
            <TouchableHighlight
              onPress={() => setScanned(false)}
              style={styles.scanButton}>
              <Text style={styles.scanButtonText}>Scan Again</Text>
            </TouchableHighlight>
          </View>
        )}
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: theme.colors.background,
  },
  scannerContainer: {
    width: 350,
    height: 560,
    borderRadius: 20,
    overflow: 'hidden',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: theme.colors.secondaryBackground,
  },
  scanner: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  permissionError: {
    fontSize: 16,
    color: theme.colors.text,
  },
  scanButtonContainer: {
    width: '100%',
    height: '100%',
    paddingHorizontal: 20,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(0,0,0, .6)',
  },
  scanButton: {
    borderRadius: 12,
    paddingVertical: 12,
    paddingHorizontal: 24,
    backgroundColor: theme.colors.background,
  },
  scanButtonText: {
    fontSize: 16,
    color: theme.colors.accent,
  },
});

export default ReadQRScreen;
