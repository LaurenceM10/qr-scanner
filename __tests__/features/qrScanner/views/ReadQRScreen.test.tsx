import React, { useEffect } from 'react';
import { Alert, View } from 'react-native';

import ReadQRScreen from 'features/qrScanner/views/ReadQRScreen';
import renderWithRedux from '../../../test-utils/renderWithRedux';

jest.mock('app/hooks/useScannerPermission', () => ({
  useScannerPermission: jest
    .fn()
    .mockReturnValueOnce(null)
    .mockReturnValueOnce(false)
    .mockReturnValue(true),
}));

function MockBarCodeScanner(
  onBarCodeScanned?: ({ data }: { data: string }) => void,
) {
  useEffect(() => {
    if (onBarCodeScanned) {
      onBarCodeScanned({ data: 'www.nicasource.com ' });
    }
  }, [onBarCodeScanned]);

  return <View />;
}

jest.mock('expo-barcode-scanner', () => ({
  BarCodeScanner: ({
    onBarCodeScanned,
  }: {
    onBarCodeScanned: ({ data }: { data: string }) => void;
  }) => MockBarCodeScanner(onBarCodeScanned),
}));

const spyAlert = jest.spyOn(Alert, 'alert');

afterEach(() => jest.clearAllMocks());

describe('<ReadQRScreen />', function () {
  it('should render the ReadQRScreen component when is requesting camera permission', function () {
    const message = 'Requesting for camera permission';
    const { toJSON, queryByText } = renderWithRedux(<ReadQRScreen />);

    expect(toJSON()).toMatchSnapshot();
    expect(queryByText(message)).toBeTruthy();
  });

  it('should render error message when there is no access to camera', function () {
    const message = 'No access to camera';
    const { queryByText } = renderWithRedux(<ReadQRScreen />);

    expect(queryByText(message)).toBeTruthy();
  });

  it('should render scanner when there is access to camera', function () {
    const { queryByTestId } = renderWithRedux(<ReadQRScreen />);

    expect(queryByTestId('ScannerContainer')).toBeTruthy();
  });

  it('should show confirmation Alert after scan element', function () {
    const { queryByTestId } = renderWithRedux(<ReadQRScreen />);

    expect(queryByTestId('ScannerContainer')).toBeTruthy();
    expect(spyAlert).toHaveBeenCalledTimes(1);
  });

  it('should save the QR item in the store if the user press the save option', function () {
    const { store } = renderWithRedux(<ReadQRScreen />);

    expect(spyAlert).toHaveBeenCalledTimes(1);
    // @ts-ignore - press the save alert option
    spyAlert.mock.calls[0][2][1].onPress();

    expect(store.getState().QRReducer.QRData.length).toBe(1);
  });
});
