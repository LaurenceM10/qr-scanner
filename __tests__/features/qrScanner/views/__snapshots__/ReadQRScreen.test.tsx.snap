// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReadQRScreen /> should render the ReadQRScreen component when is requesting camera permission 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#333333",
        "fontSize": 16,
      }
    }
  >
    Requesting for camera permission
  </Text>
</RCTSafeAreaView>
`;
